operators/pkg/instance-controller/container_logic.go

```diff
@@ -0,0 +1,206 @@
+			TCPSocket: &v1.TCPSocketAction{
+				Port: intstr.FromInt(5900),
+			},
+		},
+		InitialDelaySeconds: 3,
+		PeriodSeconds:       5,
+	}
+
+	containers := []v1.Container{
+		{
+			Name:      "novnc",
+			Image:     r.DockerGuiNovncImg + ":" + r.SvcContainersTag,
+			Resources: buildResRequirements(0.02, 50, resource.MustParse("100Mi")), // ~25MiB
+			Ports: []v1.ContainerPort{{
+				ContainerPort: httpPort,
+			}},
+			Env: []v1.EnvVar{{
+				Name:  "HIDE_NOVNC_BAR",
+				Value: strconv.FormatBool(examMode),
+			}, {
+				Name:  "HTTP_PORT",
+				Value: fmt.Sprintf("%d", httpPort),
+			}, {
+				Name:  "WEBSOCKIFY_PORT",
+				Value: fmt.Sprintf("%d", websockifyPort),
+			}},
+			SecurityContext: &contSecCtx,
+			LivenessProbe:   &noVncProbe,
+			// ReadinessProbe:  &noVncProbe,
+		},
+		{
+			Name:      "websockify",
+			Image:     r.DockerGuiWskfyImg + ":" + r.SvcContainersTag,
+			Resources: buildResRequirements(0.02, 50, resource.MustParse("50Mi")), // ~2MiB
+			Env: []v1.EnvVar{{
+				Name:  "WS_PORT",
+				Value: fmt.Sprintf("%d", websockifyPort),
+			}},
+			LivenessProbe: &websockifyProbe,
+			// ReadinessProbe: &websockifyProbe,
+		},
+		{
+			Name:            "tigervnc",
+			Image:           r.DockerGuiVncImg + ":" + r.SvcContainersTag,
+			Resources:       buildResRequirements(0.5, 50, resource.MustParse("500Mi")), // Mem depends on screen resolution, should easily manage up to 2k virtual screens
+			SecurityContext: &contSecCtx,
+			LivenessProbe:   &tigerVncProbe,
+			// ReadinessProbe:  &tigerVncProbe,
+		},
+		{
+			Name:  name,
+			Image: environment.Image,
+			Resources: buildResRequirements(
+				float32(environment.Resources.CPU),
+				environment.Resources.ReservedCPUPercentage,
+				environment.Resources.Memory,
+			),
+			SecurityContext: &contSecCtx,
+		},
+	}
+
+	template := &instance.Spec.Template
+	labels := map[string]string{
+		"name":                         name,
+		"crownlabs.polito.it/template": template.Namespace + "_" + template.Name,
+	}
+
+	return appsv1.DeploymentSpec{
+		Replicas: pointer.Int32Ptr(1),
+		Selector: &metav1.LabelSelector{
+			MatchLabels: labels,
+		},
+		Template: v1.PodTemplateSpec{
+
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: labels,
+			},
+			Spec: v1.PodSpec{
+				Containers:      containers,
+				SecurityContext: &podSecCtx,
+			},
+		},
+	}
+}
+
+// CreateContainerEnvironment implements the logic to create all the different
+// Kubernetes resources required to start a containerized CrownLabs environment.
+func (r *InstanceReconciler) CreateContainerEnvironment(
+	instance *crownlabsv1alpha2.Instance,
+	environment *crownlabsv1alpha2.Environment,
+	namespace string,
+	name string,
+	vmStart time.Time) error {
+	ctx := context.TODO()
+
+	service, ingress, err := CreateInstanceExpositionEnvironment(ctx, r, instance, name, namespace)
+	if err != nil {
+		return err
+	}
+
+	depl := appsv1.Deployment{
+		ObjectMeta: BuildInstanceObjectMeta(name, namespace, environment, instance.Name),
```

Comment:
Here you should only configure name and namespace. The labels should be set inside the mutate function